app/gamestate.ts

```diff
@@ -0,0 +1,390 @@
+import * as wire from './wire';
+import RobotProxy from './robotproxy';
+import { RobotGameObject, RegenGameObject } from './gameobjects';
+
+function randomInteger(max: number, min = 0) {
+    return Math.floor(Math.random() * (max - min + 1)) + min;
+}
+
+function directionToLoaction(location: wire.Point, direction: wire.CardinalDirection) {
+    const newLoc = {x: location.x, y: location.y};
+    switch (direction) {
+        case wire.CardinalDirection.East:
+            newLoc.x += 1;
+        case wire.CardinalDirection.South:
+            newLoc.y -= 1;
+        case wire.CardinalDirection.West:
+            newLoc.x -= 1;
+        case wire.CardinalDirection.North:
+            newLoc.y += 1;
+    }
+    return newLoc;
+}
+
+export class GameState {
+    public entities: {
+        robots: RobotGameObject[],
+        regens: RegenGameObject[],
+    };
+    private grid: {[index: string]: boolean};
+    constructor(public width: number, public height: number, robots: {code: string}[]) {
+        this.entities = {
+            robots: [],
+            regens: [],
+        };
+
+        this.grid = {};
+        let counter = 0;
+        // TODO change to map
+        for (const robot of robots) {
+            // find a location where we can put our robot
+            const robotLocation = this.getAvailableLocation();
+            this.markLocation(robotLocation);
+
+            // create our robot with that location
+            let robotId = (counter++).toString();
+            this.entities.robots.push( new RobotGameObject(robotId, robot.code,
+                wire.Health, wire.Costs.income, robotLocation) );
+        }
+
+        this.spawnRegens(robots.length - 1);
+
+        this.entities.robots.forEach(robot => robot.setupProxy(this.entities.robots.map(robot2 => this.initialInfoForOf(robot, robot2)), {x: width, y: height}));
+
+        // TODO: Generate correct starting data, loop until match over
+        Promise.all(
+            this.entities.robots.map(robot => robot.proxy.getTurn(this.getStateFor(robot)))
+        ).then(listOfActionMessages => this.simulateTurn(listOfActionMessages));
+    }
+
+    getAvailableLocation(): wire.Point {
+        let placeX: number, placeY: number;
+        let placeLocation = '';
+        do {
+            placeX = randomInteger(this.width);
+            placeY = randomInteger(this.height);
+            placeLocation = `${placeX},${placeY}`;
+        } while (this.grid[placeLocation] === undefined);
+        return {x: placeX, y: placeY};
+    }
+
+    markLocation(point: wire.Point) {
+        this.grid[`${point.x},${point.y}`] = true;
```

Comment:
Point is an interface, not a class - any is object with a numeric x and a y satisfies the interface. (Plus, the distinction is good at reminding you that the value of the map needs to contain the point object if you want to keep it easily accessible from the map).

I have, however, considered writing a `toKey` helper a few times. Its probably a good idea.